utility_code/lib_py3/plugin_data.py

```diff
@@ -0,0 +1,295 @@
+
+    yield from process_in_parallel(parallel_args, num_processes=num_processes, initializer=initializer, initargs=initargs)
+
+class PluginData(NbtPathDebug):
+    """Plugin data loaded alongside player data."""
+
+    def __init__(self, json_data=None, json_path=None):
+        """Load a plugin data from json.
+
+        Must be saved from wherever the json was loaded from for changes to apply.
+        """
+        self._path = json_path
+        self._data = json_data
+
+        if self._path is not None and self._data is None:
+            with open(self._path, 'r') as fp:
+                self._data = json.load(fp)
+                fp.close()
+
+        if self._data is None:
+            raise ValueError("No json data or path was provided")
+
+        self.nbt_path_init(None, None, self, None)
+
+    def get_debug_str(self):
+        return str(self)
+
+    def save(self, json_path=None):
+        """Save json data.
+
+        Always returns the json data.
+        If loaded from a path, or a path is provided, saves to that path.
+        """
+        if json_path is None:
+            json_path = self._path
+
+        if json_path:
+            with open(json_path, 'w') as fp:
+                json.dump(
+                    self._data,
+                    fp,
+                    ensure_ascii=False,
+                    indent=2,
+                    separators=(',', ': '),
+                    sort_keys=False
+                )
+                fp.write("\n")
+
+        return self._data
+
+    def graves(self):
+        """Get the graves stored on a player, if they exist."""
+        return MonumentaGraves(self._data.get("MonumentaGravesV2", {}), self)
+
+    def __repr__(self):
+        return f'PluginData({os.path.basename(self._path)!r})'
+
+class MonumentaGraves(NbtPathDebug):
+    """A collection of items preserved from death or carelessness."""
+
+    def __init__(self, json_data, parent):
+        self._data = json_data
+
+        self.nbt_path_init(None, parent, parent.root if parent is not None and parent.root is not None else self, None)
+
+    def get_debug_str(self):
+        return str(self)
+
+    def iter_graves(self):
+        graves = self._data.get("graves", [])
+        num_graves = len(graves)
+        for i in range(num_graves):
+            grave = MonumentaGrave(graves[i], self)
+            yield grave
+            graves[i] = grave.serialize()
+
+    def iter_thrown_items(self):
+        thrown_items = self._data.get("thrown_items", [])
+        num_thrown_items = len(thrown_items)
+        for i in range(num_thrown_items):
+            thrown_item = JsonWrappedItem(thrown_items[i], self)
+            yield thrown_item
+            thrown_items[i] = thrown_item.serialize()
+
+    def iter_all_types(self):
+        yield from self.iter_graves()
+        yield from self.iter_thrown_items()
+
+    def iter_block_entities(self):
+        pass
```

Comment:
These three `pass` statements should be `return` with no return value. This is functionally the same, more of a style thing.

Reply:
Will fix